File: renderer/src/interfaces/apiData.ts
import { HttpMethod, Https } from "../enums/Apis/method";
import { FormDataItem, RequestDataType, TimeOutType } from "../enums/VariableEnum";
import { KeyValueCheckRecord } from "./variableData";

export interface Cookie {
    key: string;
    value: string;
    expires: Date | 'Infinity' | null;
    maxAge: number | 'Infinity' | '-Infinity' | null;
    domain: string | null;
    path: string | null;
    secure: boolean;
    httpOnly: boolean;
    extensions: string[] | null;
    creation: Date | 'Infinity' | null;
    creationIndex: number;
    hostOnly: boolean | null;
    pathIsDefault: boolean | null;
    lastAccessed: Date | 'Infinity' | null;
    sameSite: string | undefined;
}


export interface FormDataType {
    key: string,
    value: string,
    enabled: boolean,
    type: FormDataItem
}

export interface ApiData {
    nonce: string,
    rawCode: string,
    url: string,
    name: string;
    method: HttpMethod
    https: Https
    params: KeyValueCheckRecord[],
    headers: KeyValueCheckRecord[],
    path: Record<string, string>,
    requestCookies: Cookie[],
    timeout: number,
    timeoutType: TimeOutType,
    requestDataType: RequestDataType,
    tests: {
        name: string,
        function: string
    }[],
    json: string | undefined,
    xml: string | undefined,
    js: string | undefined,
    html: string | undefined,
    text: string | undefined,
    formData: FormDataType[],
    urlEncoded: KeyValueCheckRecord[],
    binary: string | undefined,
    examples: {
        name: string,
        path: string,
    }[]
}

export interface ApiResponse {
    invoked: boolean,
    name: string;
    saved: boolean,
    error: boolean;
    mime: string;
    parsedUrl: string;
    timeTaken: number;
    data: any;
    status: number;
    statusText: string;
    headers: Object;
    size: number;
    cookie: Cookie[];
    errorMessage: string[],
    warningMessage: string[],
    varUsed: Record<string, string>;
    tests: {
        name: string,
        status: boolean,
        message: string
    }[]
}

File: src/main/kotlin/com/aayam/toasteditor/constants/interfaces/apis/ApiData.kt
package com.aayam.toasteditor.constants.interfaces.apis

import com.aayam.toasteditor.constants.enums.HttpMethod
import com.aayam.toasteditor.constants.enums.Https
import com.aayam.toasteditor.constants.enums.RequestDataType
import com.aayam.toasteditor.constants.enums.TimeOutType
import com.aayam.toasteditor.constants.interfaces.datas.KeyValueCheckRecord
import kotlinx.serialization.Serializable

@Serializable
data class ApiData(
    val nonce: String,
    val rawCode: String,
    val url: String,
    val name: String,
    val method: HttpMethod,
    val https: Https,
    val params: List<KeyValueCheckRecord> = emptyList(),
    val headers: List<KeyValueCheckRecord> = emptyList(),
    val path: Map<String, String> = mapOf(),
    val requestCookies: List<Cookies> = emptyList(),
    val timeout: Float,
    val timeoutType: TimeOutType,
    val requestDataType: RequestDataType,
    val tests: List<ApiDataTest> = emptyList(),
    val json: String? = null,
    val xml: String? = null,
    val js: String? = null,
    val html: String? = null,
    val text: String? =  null,
    val formData: List<FormDataType> = emptyList(),
    val urlEncoded: List<KeyValueCheckRecord> = emptyList(),
    val binary: String? = null,
    val examples: List<ApiDataExample> = emptyList()
)
